docs/decisions/001X-python-client-constructors.md

```diff
@@ -0,0 +1,165 @@
+---
+status: proposed
+contact: eavanvalkenburg
+date: 2025-11-18
+deciders: markwallace-microsoft, dmytrostruk, sphenry, alliscode
+consulted: taochenosu, moonbox3, giles17
+---
+
+# Python Client Constructors
+
+## Context and Problem Statement
+
+We have multiple Chat Client implementations that can be used with different servers, the most important example is OpenAI, where we have a separate client for OpenAI and for Azure OpenAI. The constructors for the underlying OpenAI client has now enabled both, so it might make sense to have a single Chat Client for both, the same also applies to other Chat Clients, such as Anthropic, which has a Anthropic client, but also AnthropicBedrock and AnthropicVertex, currently we don't support creating a AF AnthropicClient with those by default, but if you pass them in as a parameter, it will work. This is not the case for OpenAI, where we have a separate client for Azure OpenAI and OpenAI, the OpenAI clients still accept any OpenAI Client (or a subclass thereof) as a parameter, so it can already be used with different servers, including Azure OpenAI, but it is not the default.
+
+We have a preference of creating clients inside of our code because then we can add a user-agent string to allow us to track usage of our clients with different services. This is most useful for Azure, but could also be a strong signal for other vendors to invest in first party support for Agent Framework. And we also make sure to not alter clients that are passed in, as that is often meant for a specific use case, such as setting up httpx clients with proxies, or other customizations that are not relevant for the Agent Framework.
+
+There is likely not a single best solution, the goal here is consistency across clients, and with that ease of use for users of Agent Framework.
+
+## Decision Drivers
+
+- Reduce client sprawl and different clients that only have one or more different parameters.
+- Make client creation inside our classes the default and cover as many backends as possible.
+- Make clients easy to use and discover, so that users can easily find the right client for their use case.
+- Allow client creation based on environment variables, so that users can easily configure their clients without having to pass in parameters.
+
+## Considered Options
+
+- Separate clients for each backend, such as OpenAI and Azure OpenAI, Anthropic and AnthropicBedrock, etc.
```

Comment:
Our current design uses inheritance (`AzureOpenAIChatClient` extends `OpenAIBaseChatClient`). Would any of the proposed options require shifting to composition patterns instead? How would that impact the `@use_function_invocation`, `@use_observability`, and `@use_chat_middleware` decorators?